components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -14,9 +18,78 @@ pub enum CanonicalizationResult {
 }
 
 pub struct LocaleCanonicalizer<'a> {
+    aliases: DataPayload<'a, AliasesV1>,
     likely_subtags: DataPayload<'a, LikelySubtagsV1>,
 }
 
+#[inline]
+fn uts35_rule_matches(source: &Locale, ruletype: &LanguageIdentifier) -> bool {
+    (ruletype.language.is_empty() || ruletype.language == source.id.language)
+        && (ruletype.script.is_none() || ruletype.script == source.id.script)
+        && (ruletype.region.is_none() || ruletype.region == source.id.region)
+        && (ruletype.variants.is_empty()
+            || ruletype
+                .variants
+                .iter()
+                .all(|v| source.id.variants.contains(v)))
```

Comment:
`question`: that leaves out a scenario where `source` has additional variants not present in `ruletype` - is that okay?

Reply:
Yes, UTS-35 says to match if the ruletype variants are a subset of the source variants, so `ja-Latn-fonipa-hepburn-heploc` matches against the rule for `hepburn-heploc` and is canonicalized to `ja-Latn-alalc97-fonipa`. I'll add a comment explaining this is intentional.